### DISCUSSION TITLE
How to call Discussion APIs in Sunbird Lern BB from InQuiry BB?

### DISCUSSION THREAD

**Discussion by joffinjoy** [Sunbird-inQuiry/Community#22] [Q&A]
@maheshkumargangula @vrayulu @pramodkvarma
With regards to the reviewer flow [discussion](https://github.com/orgs/Sunbird-inQuiry/discussions/8#discussioncomment-3145136), we were suggested to make use of discussion APIs by leveraging Sunbird Lern BB. So how can we access Lern APIs from Inquiry in various environments? What is the suggested workflow for interaction between various BBs?

Current idea is to create a new **Category** in nodeBB for all reviewer comments. And then have one **topic** per questionSet that is being reviewed. Each **comment** by the reviewer would be added as a **post** to the respective topic. We would add fields to both question and questionSet objects to store the id that uniquely identities its topic and post and fetch it while displaying the questionset to reviewer or creator.

And do we have docker setup documentation for nodeBB and related dependencies? 

cc: @aks30, @vijiurs, @kiranharidas187 @aishwaryashikshalokam, @priyanka-TL, @VISHNUDAS-tunerlabs, @Shakthieshwari, @Prateek-slokam, @MeghnaSL, @joffinjoy, @githubpiyush

**Comment by pramodkvarma** [comment]
Just curious! Why is inQuiry handling comments? Shouldn't that be part of Lern or coKreat?

**Reply by aks30** [reply]
@pramodkvarma For the discussion here - https://github.com/orgs/Sunbird-inQuiry/discussions/8#discussioncomment-3145136 where we said we can leverage on the **Discussion Forum APIs** to store Reviewer Comments and thereafter the creator reply on it  in terms of posts and comments on posts in Discussion Forum, for that the InQuiry BB (which is the BB for Question/Question set service and the editors enabled on creation portal) will need to talk to NodeBB (nodebb-service) which is available in Sunbird Lern BB. So what @joffinjoy was trying to ask is, whether or not in the VDN setup (https://dock.sunbirded.org/) the NodeBB service is also deployed so as to integrate it for review comments workflow in the Question set creation lifecycle.

**Reply by pramodkvarma** [reply]
I understood that part. My question remains. Why does inQuiry BB dealing with questions/sets dealing with comments? Isn't the creation/management flow part of coKreat? If so, coKreat should ideally stitch it together, right?

**Reply by aks30** [reply]
@pramodkvarma We could be wrong, but our plan was that the comments will be added and viewed using editor code, and to store topic id and post ids against question-set and questions we will be using the Question and Question set service which is being described here in the InQuiry BB - https://inquiry.sunbird.org/learn/product-and-developer-guide/question-and-question-set-service

<img width="832" alt="Screenshot 2022-07-15 at 6 27 59 PM" src="https://user-images.githubusercontent.com/9779779/179227616-a70e5613-ab33-45c6-b1fb-14dc19d40a41.png">
 
Based on this we thought it of it to be part of Inquiry BB.

**Reply by vrayulu** [reply]
@aks30 - though inQuiry has to provide the APIs for lifecycle events of questions & question sets, it need not have the APIs for the supporting activities like review discussions, etc. This is mainly for adhering to the principle of unbundling and keeping the micro-services minimal. An example of it is how course discussions are enabled in SunbirdEd - SunbirdEd uses both course APIs and discussion APIs to enable a combined experience of courses + discussions to the users. There is not much value in Course APIs (or inQuiry APIs in the current scenario) providing a wrapper on discussion forum APIs to enable discussions on a course (or a question set). Discussion APIs are flexible and extensible enough to support discussions in multiple different contexts.

Another example of such unbundling is how coKreat enabled contribution functionality using contribution APIs on top of inQuiry APIs. This unbundling is important for imagination & creation of a wide range of solutions by integrating multiple building blocks.

**Reply by pramodkvarma** [reply]
Great point @vrayulu. Agree with the idea of Lern or coKreat stitching together the functionality using multiple APIs rather than creating a monolithic function at inQuiry BB level.

**Reply by aks30** [reply]
Totally agree @vrayulu and @pramodkvarma! The clarity and guidance we wanted via this discussion forum post is how do we integrate/call/leverage/use Discussion forum APIs for storing and rendering back the reviewer comments during the Review phase of Question-Set creation and publishing as in, are the Topic/Post related APIs from Discussion Forum service available in VDN setup at present? As per the diagram we see - https://cokreat.sunbird.org/learn/technical-architecture it's not. If it is, if someone can guide us on how to do that and if not how should we bring that into the VDN setup?

@joffinjoy I suggest based on @vrayulu 's comment we should move this discussion to the CoKreat BB forum.

**Reply by pramodkvarma** [reply]
Yes, definitely, we should enable coKreat and  Lern to leverage discussions API for comments on courses, during creation, etc.

**Reply by vrayulu** [reply]
@aks30 - from an implementation point of view, the reviewer comments functionality should be able to work with an instance of Sunbird discussions service. The discussions service instance could be part of coKreat or sunbird-ed or a stand-alone discussion service instance.  

If the environment in which the feature is deployed already has a running instance of discussion service, this feature should be configured to use that instance.
